### PR TITLE
new resource: `azurerm_data_factory_credential_service_principal`

### PR DESCRIPTION
**Pull request by ms-henglu**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave "+1" or "me too" comments, they generate extra noise for PR followers and do not help prioritize for review

## PR Checklist

- [ ] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [ ] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [ ] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [ ] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [ ] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [ ] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [ ] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [ ] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [ ] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

image

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_resource` - support for the `thing1` property [GH-00000]

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #0000

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: internal/services/datafactory/data_factory_credential_service_principal_resource.go [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+			}
+
+			id := credentials.NewCredentialID(dataFactoryId.SubscriptionId, dataFactoryId.ResourceGroupName, dataFactoryId.FactoryName, data.Name)
+			existing, err := client.CredentialOperationsGet(ctx, id, credentials.DefaultCredentialOperationsGetOperationOptions())
+			if err != nil && !response.WasNotFound(existing.HttpResponse) {
+				return fmt.Errorf("checking for presence of existing %s: %+v", id, err)
+			}
+
+			if !response.WasNotFound(existing.HttpResponse) {
+				return tf.ImportAsExistsError("azurerm_data_factory_dataset_http", id.ID())
```
Comment (magodo) [inline internal/services/datafactory/data_factory_credential_service_principal_resource.go:218]: Wrong RT?

File: internal/services/datafactory/data_factory_credential_service_principal_resource.go [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+				return fmt.Errorf("checking for presence of existing %s: %+v", id.ID(), err)
+			}
+
+			if existing.Model == nil {
+				return fmt.Errorf("retrieving %s: `model` was nil", *id)
+			}
+
+			props, ok := existing.Model.Properties.(credentials.ServicePrincipalCredential)
+			if !ok {
+				return fmt.Errorf("retrieving %s: expected `credentials.ServicePrincipalCredential` but got %+v", id, existing.Model.Properties)
```
Comment (magodo) [inline internal/services/datafactory/data_factory_credential_service_principal_resource.go:282]:
```suggestion
				return fmt.Errorf("retrieving %s: expected `credentials.ServicePrincipalCredential` but got %+T", id, existing.Model.Properties)
```

**Review by magodo — CHANGES_REQUESTED** [review]
Thanks for this PR!
I've taken a look through and left some comments inline, but this is mostly looking good to me 👍

File: internal/services/datafactory/data_factory_credential_service_principal_resource.go [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+			ValidateFunc: validation.IsUUID,
+		},
+
+		"service_principal_key": {
+			Type:     pluginsdk.TypeList,
+			Optional: true,
+			MaxItems: 1,
+			Elem: &pluginsdk.Resource{
+				Schema: map[string]*pluginsdk.Schema{
+					"linked_service_name": {
```
[the remark this answers is not in the document]
Comment (ms-henglu) [reply internal/services/datafactory/data_factory_credential_service_principal_resource.go:83]: Yes,

[the remark this answers is not in the document]
Comment (magodo) [reply internal/services/datafactory/data_factory_credential_service_principal_resource.go:83]: Then I feel a single key vault secret id suffices here?

[the remark this answers is not in the document]
Comment (ms-henglu) [reply internal/services/datafactory/data_factory_credential_service_principal_resource.go:83]: No, the linked service name here is the name of a data factory linked service which connects to the key vault.. So the name could be different from the key vault name.

**Comment by magodo** [comment]
LGTM now!

**Comment by ms-henglu** [comment]
Thanks @magodo

**Review by katbyte — CHANGES_REQUESTED** [review]
Test failure:
```
------- Stdout: -------
=== RUN   TestAccDataFactoryCredentialServicePrincipal_requiresImport
=== PAUSE TestAccDataFactoryCredentialServicePrincipal_requiresImport
=== CONT  TestAccDataFactoryCredentialServicePrincipal_requiresImport
    testcase.go:113: Step 1/2 error: Error running apply: exit status 1
        
        Error: creating Key Vault (Subscription: "*******"
        Resource Group Name: "acctestRG-df-240506171043463502"
        Key Vault Name: "acctestkv24050602"): performing CreateOrUpdate: vaults.VaultsClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=<nil> Code="VaultAlreadyExists" Message="The vault name 'acctestkv24050602' is already in use. Vault names are globally unique so it is possible that the name is already taken. If you are sure that the vault name was not taken then it is possible that a vault with the same name was recently deleted but not purged after being placed in a recoverable state. If the vault is in a recoverable state then the vault will need to be purged before reusing the name. For more information about VaultAlreadyExists, soft delete and purging a vault follow this link https://go.microsoft.com/fwlink/?linkid=2147740."
        
          with azurerm_key_vault.test,
          on terraform_plugin_test.tf line 44, in resource "azurerm_key_vault" "test":
          44: resource "azurerm_key_vault" "test" {
        
--- FAIL: TestAccDataFactoryCredentialServicePrincipal_requiresImport (66.25s)
FAIL
```

**Comment by ms-henglu** [comment]
Thanks @katbyte , I've updated the tests, now all passing

image

**Review by katbyte — APPROVED** [review]
LGTM 🌈

**Pull request merged**